### PR TITLE
[python/r] Use pushdown domainish accessors at the Python/R UX level

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:** As tracked on issue #2407 / [[sc-51048]](https://app.shortcut.com/tiledb-inc/story/51048).

See also [[sc-55659]](https://app.shortcut.com/tiledb-inc/story/55659/r-current-domain-tasks).

Note that the intended Python and R API changes are all agreed on and finalized as described in #2407.

**Changes:**

Following #3012 we can now connect tiledbsoma-py to use the full domainish accessors from `libtiledbsoma`.

**Notes for Reviewer:**

R mods to come on this same PR.

This PR is a work in progress. It is not ready for review.

### CODE REVIEW

File: apis/r/R/SOMADataFrame.R
Comment (mojaveazure) [inline apis/r/R/SOMADataFrame.R:351]: What's the R-object return type of `domain()` (and `maxdomain()` and `non_empty_domain()`)? Why do we take the results and go to an Arrow Table, then an Arrow RecordBatch, then a list? Do we need these conversions in the first place? I know that you can call `as.list()` on an Arrow Table directly, but do we need to go through a Table in the first place?

```R
> (tbl <- arrow::as_arrow_table(mtcars))
Table
32 rows x 11 columns
$mpg <double>
$cyl <double>
$disp <double>
$hp <double>
$drat <double>
$wt <double>
$qsec <double>
$vs <double>
$am <double>
$gear <double>
$carb <double>

See $metadata for additional Schema metadata
> identical(as.list(tbl), as.list(arrow::as_record_batch(tbl)))
[1] TRUE
```

File: apis/r/R/utils-arrow.R
Comment (mojaveazure) [inline apis/r/R/utils-arrow.R:381]: Style fix
```suggestion
          ind_max_dom <- arrow_type_unsigned_range(ind_col_type) - c(0, ind_ext)
```

File: apis/r/R/utils-arrow.R
Comment (mojaveazure) [inline apis/r/R/utils-arrow.R:384]: Style fix
```suggestion
          ind_max_dom <- arrow_type_range(ind_col_type) - c(0, ind_ext)
```

File: apis/r/tests/testthat/test-shape.R
Comment (mojaveazure) [inline apis/r/tests/testthat/test-shape.R:10]: These test modifications can be simplified:
 - the names of the list aren't being used for anything essential
```suggestion
    "soma_joinid",
    c("soma_joinid", "foo"),
    c("soma_joinid", "baz"),
    c("baz", "foo")
```

File: apis/r/tests/testthat/test-shape.R
Comment (mojaveazure) [inline apis/r/tests/testthat/test-shape.R:14]: These test modifications can be simplified:
 - we're iterating through a list and performing the same sets of checks, we can use `seq_along()` to iterate along the list for us
```suggestion
  for (i in seq_along(index_column_name_choices)) {
    index_column_names <- index_column_name_choices[[i]]
```

File: apis/r/tests/testthat/test-shape.R
Comment (mojaveazure) [inline apis/r/tests/testthat/test-shape.R:63]: Style fixes
```suggestion
    dom <- sdf$domain()
    mxd <- sdf$maxdomain()
```

File: apis/r/tests/testthat/test-shape.R
Comment (mojaveazure) [inline apis/r/tests/testthat/test-shape.R:81]: These test modifications can be simplified:
 - we're checking that the names of `dom` and `mxd` are the same as `index_column_names` for each case; rather than hard-coding the cases, we can use the already-defined variable to perform the check for us. This also allows us to add new cases should we desire
```suggestion
    # First check names
    expect_equal(names(dom), index_column_names)
    expect_equal(names(mxd), index_column_names)
```

File: apis/r/tests/testthat/test-shape.R
Comment (mojaveazure) [inline apis/r/tests/testthat/test-shape.R:89]: These test modifications can be simplified:
 - we've already determined that the names of `dom` and `mxd` are identical; we can iterate through them both simultaneously
 - we can use `expec_length()` instead of `expect_equal()` for more informative test failure messages
```suggestion
    for (name in names(dom)) {
      expect_length(dom[[name]], 2L)
      expect_length(mxd[[name]], 2L)
    }
```

**Review by mojaveazure — CHANGES_REQUESTED** [review]

File: apis/r/R/SOMADataFrame.R
Comment (johnkerl) [reply apis/r/R/SOMADataFrame.R:351]: I found that these steps were all necessary. Errors otherwise. I can share details if you like.

Comment (johnkerl) [reply apis/r/R/SOMADataFrame.R:351]: The R object returned is a named list of two-element vectors.

Comment (mojaveazure) [reply apis/r/R/SOMADataFrame.R:351]: Please do share details; it doesn't make sense for a list-return to throw errors or for `as.list(tbl)` to fail while `as.list(arrow::as_record_batch(tbl))` works

Comment (johnkerl) [reply apis/r/R/SOMADataFrame.R:351]: @mojaveazure 

```
# This is a local copy of pbmc3k
> sdf <- SOMAOpen('/var/p/obs')

> sdf$schema()
Schema
soma_joinid: int64 not null
obs_id: string not null
n_genes: int64 not null
percent_mito: float not null
n_counts: float not null
louvain: dictionary<values=string, indices=int8> not null

> ctx <- tiledbsoma:::soma_context()

> a <- domain(sdf$uri, ctx)

> a
<nanoarrow_array struct[0]>
 $ length    : int 0
 $ null_count: int 0
 $ offset    : int 0
 $ buffers   :List of 1
  ..$ :<nanoarrow_buffer validity<bool>[0][0 b]> ``
 $ children  :List of 1
  ..$ soma_joinid:<nanoarrow_array int64[2]>
  .. ..$ length    : int 2
  .. ..$ null_count: int 0
  .. ..$ offset    : int 0
  .. ..$ buffers   :List of 2
  .. .. ..$ :<nanoarrow_buffer validity<bool>[0][0 b]> ``
  .. .. ..$ :<nanoarrow_buffer data<int64>[2][16 b]> `0.00e+00 2.15e+09`
  .. ..$ dictionary: NULL
  .. ..$ children  : list()
 $ dictionary: NULL

> as.list(a)
Error in as.vector.nanoarrow_array(x, "list") :
  identical(mode, "any") is not TRUE

> as.list(arrow::as_record_batch(a))
Error in `[<-`:
! Assigned data `map(.subset(x, unname), vectbl_set_names, NULL)` must be compatible with existing data.
✖ Existing data has 0 rows.
✖ Assigned data has 2 rows.
ℹ Only vectors of size 1 are recycled.
Caused by error in `vectbl_recycle_rhs_rows()`:
! Can't recycle input of size 2 to size 0.
Run `rlang::last_trace()` to see where the error occurred.

> as.list(arrow::as_arrow_table(a))
Error in `[<-`:
! Assigned data `map(.subset(x, unname), vectbl_set_names, NULL)` must be compatible with existing data.
✖ Existing data has 0 rows.
✖ Assigned data has 2 rows.
ℹ Only vectors of size 1 are recycled.
Caused by error in `vectbl_recycle_rhs_rows()`:
! Can't recycle input of size 2 to size 0.
Run `rlang::last_trace()` to see where the error occurred.

> as.list(arrow::as_record_batch(arrow::as_arrow_table(a)))
$soma_joinid
[1]          0 2147483646
```

Comment (mojaveazure) [reply apis/r/R/SOMADataFrame.R:351]: Ah, okay. It seems like `domain()` et al don't actually read in the domain but leave the domain information in a nanoarrow buffer. I don't know if going from nanoarrow array w/ buffer -> Arrow Table -> Arrow RecordBatch -> list is the best way forward, but I can't find a way around it in R (especially since it seems like the RecordBatch is needed to get the data out of the buffer)

Comment (mojaveazure) [reply apis/r/R/SOMADataFrame.R:351]: My question now is do we want these returns to be a list or a vector? Another way of putting that is will these functions always return two values (min, max) or potentially multiple sets of min/max pairs?

Comment (johnkerl) [reply apis/r/R/SOMADataFrame.R:351]: @mojaveazure that's partly a UX question

* Each slot's _value_ will always be a 2-element lo/hi pair -- one per dim -- no changing that
* What's a UX questiion is the _keys_ -- what's better, a named list (which I personally like), or, an unnamed list, or, an outer vector of length N where N is the number of dims

I personally like the above three choices in the order I listed them, b/c (a) there's more info there; (b) people can remove info by `unname(dom)` or `as.vector(x)` but it's harder to put the info back in if we just give them a length-N vector of length-2 vectors

But that's just my general feel as a polyglottal language user; I defer to you

I'd also point out that on the Python side it's a length-n tuple of length-2 tuples ... so for strict partity we might go that way ...

Comment (johnkerl) [reply apis/r/R/SOMADataFrame.R:351]: Maybe have an optional `as.something` arg to the domainish accessors where the default is the compact way but people can get the named-list info if they ask for it? Just a thought. (And it doesn't add much complexity.)

Comment (mojaveazure) [reply apis/r/R/SOMADataFrame.R:351]: Vectors are strictly flat and cannot be nested. Because it seems like the base unit of return is the dim and there's effectively one return (a min/max pair) per dim, a list is the appropriate return. I agree that named is better than unnamed as it's easer to unname than to re-name

Comment (mojaveazure) [reply apis/r/R/SOMADataFrame.R:351]: (I apologize for the back-and-forth on this; I was not part of the discussions for these new methods, so I'm learning them for the first time here)

**Review by mojaveazure — APPROVED** [review]
LGTM on the R-side; will defer to other for Python. Before merging, please bump the develop version and update the changelog

File: apis/python/src/tiledbsoma/_tdb_handles.py
Comment (nguyenv) [inline apis/python/src/tiledbsoma/_tdb_handles.py:366]: Is the `domainish` type not a `Tuple[Tuple[object, object], ...]`?

Comment (nguyenv) [inline apis/python/src/tiledbsoma/_tdb_handles.py:366]: Would the type for `domainish` be something like `List[np.array[object, object], ...]`?

File: apis/python/src/tiledbsoma/_tdb_handles.py
Comment (nguyenv) [inline apis/python/src/tiledbsoma/_tdb_handles.py:369]: I might have missed it, but I don't think you actually use `i` in this loop.

```suggestion
        for slot in domainish:
```

**Review by nguyenv — APPROVED** [review]

File: apis/python/src/tiledbsoma/_tdb_handles.py
Comment (johnkerl) [reply apis/python/src/tiledbsoma/_tdb_handles.py:369]:
```
    index_column_names <- index_column_name_choices[[i]]
```

Comment (johnkerl) [reply apis/python/src/tiledbsoma/_tdb_handles.py:369]: and on #3032, more there

File: apis/python/src/tiledbsoma/_tdb_handles.py
Comment (johnkerl) [reply apis/python/src/tiledbsoma/_tdb_handles.py:366]: yeah something like that

it gave me fits & I did `Any` to unblock & forgot to come back & fine-tune ... let me try that ...

thank you! :)

Comment (johnkerl) [reply apis/python/src/tiledbsoma/_tdb_handles.py:366]: @nguyenv I merged without doing this (I resolved the wrong comment) -- will address in follow-up

Comment (johnkerl) [reply apis/python/src/tiledbsoma/_tdb_handles.py:366]: https://github.com/single-cell-data/TileDB-SOMA/issues/3081

Comment (johnkerl) [reply apis/python/src/tiledbsoma/_tdb_handles.py:366]: @nguyenv I may leave it as `Any` but with a comment. Here's what it looks like in one example:

```
>>> sdf._handle._handle.domain()
[<pyarrow.lib.LargeStringArray object at 0x100d8c6a0>
[
  "apple",
  "zebra"
], <pyarrow.lib.Int32Array object at 0x1612e5e40>
[
  -1000000,
  1000000
]]
>>> sdf._handle._handle.domain()[0]
<pyarrow.lib.LargeStringArray object at 0x1612e5ea0>
[
  "apple",
  "zebra"
]
>>> sdf._handle._handle.domain()[1]
<pyarrow.lib.Int32Array object at 0x1612e5e40>
[
  -1000000,
  1000000
]
```

So it's a list of `pyarrow` arrays ...

**Pull request merged**
